language/stdlib/src/transaction_scripts.rs

```diff
@@ -1,74 +1,210 @@
+    CreateAccount,
+    EmptyScript,
+    Mint,
+    PeerToPeer,
+    PeerToPeerWithMetadata,
+    PlaceholderScript,
+    RegisterValidator,
+    RemoveValidator,
+    RotateAuthenticationKey,
+    RotateConsensusPubkey,
+    // ...add new scripts here
 }
 
-fn from_staged(script_name: &str) -> Vec<u8> {
-    let mut path = PathBuf::from(script_name);
-    path.set_extension(STAGED_EXTENSION);
-    STAGED_TXN_SCRIPTS_DIR
-        .get_file(path)
-        .unwrap()
-        .contents()
-        .to_vec()
-}
+impl StdlibScript {
+    // TODO: audit/cleanup. There is some cruft here; e.g., PlaceholderScript and EmptyScript are
+    // the same.
+    /// Return a vector containing all of the standard library scripts (i.e., all inhabitants of the
+    /// StdlibScript enum)
+    pub fn all() -> Vec<Self> {
+        vec![
+            Self::AddValidator,
+            Self::CreateAccount,
+            Self::EmptyScript,
+            Self::Mint,
+            Self::PeerToPeer,
+            Self::PeerToPeerWithMetadata,
+            Self::PlaceholderScript,
+            Self::RegisterValidator,
+            Self::RemoveValidator,
+            Self::RotateAuthenticationKey,
+            Self::RotateConsensusPubkey,
+            // ...add new scripts here
+        ]
+    }
 
-// Prepends the directory for the transaction scripts, adds the move file suffix to the given
-// source file locatio, and then compiles this.
-fn from_current_source(script_name: &str) -> Vec<u8> {
-    let mut path = TXN_SCRIPTS_DIR.clone();
-    path.push(PathBuf::from(script_name));
-    path.set_extension(MOVE_EXTENSION);
-    let final_path = path.into_os_string().into_string().unwrap();
-    compile_script(final_path)
-}
+    /// Construct the whitelist of script hashes used to determine whether a transaction script can
+    /// be executed on the Libra blockchain
+    pub fn whitelist() -> Vec<[u8; SCRIPT_HASH_LENGTH]> {
+        StdlibScript::all()
+            .iter()
+            .map(|script| *script.compiled_bytes().hash().as_ref())
+            .collect()
+    }
 
-pub static ADD_VALIDATOR_TXN: Lazy<Vec<u8>> = Lazy::new(|| compiled_script("add_validator"));
+    /// Return a lowercase-underscore style name for this script
+    pub fn name(self) -> String {
+        self.to_string()
+    }
 
-pub static PEER_TO_PEER_TRANSFER_TXN: Lazy<Vec<u8>> =
-    Lazy::new(|| compiled_script("peer_to_peer_transfer"));
+    /// Return true if `code_bytes` is the bytecode of one of the standard library scripts
+    pub fn is(code_bytes: &[u8]) -> bool {
+        Self::try_from(code_bytes).is_ok()
+    }
 
-pub static PEER_TO_PEER_TRANSFER_WITH_METADATA_TXN: Lazy<Vec<u8>> =
-    Lazy::new(|| compiled_script("peer_to_peer_transfer_with_metadata"));
+    /// Return the Move bytecode produced by compiling this script. This will almost always hit the
+    /// cache rather than invoking the compiler; genesis is the only exception.
+    pub fn compiled_bytes(self) -> CompiledBytes {
```

Comment:
I wonder about this clone() and returning the vector itself here but I understand we are compiling it and may get difficult to  spend time trying to make this "lighter".
I wondered because I think some usages (like `hash()`?) use the reference and so we may make a copy for nothing.
In any case not a blocker at all as this looks like an improvement to me

Reply:
> returning the vector itself here

Agreed that this could make sense--I was very on the fence about whether the `CompiledBytes`  abstraction is overkill or not.

> I wondered because I think some usages (like hash()?) use the reference and so we may make a copy for nothing.

I think we unfortunately can't return a reference because the `else` branch builds a vector from scratch.